### PR TITLE
Fixed #25393 -- Don't fail MySQL add of TextField with unhashable default

### PR DESCRIPTION
**Pull request by scop**
https://code.djangoproject.com/ticket/25393

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by MarkusH** [comment]
Since the tests seem to pass on all database backends I guess the others don't suffer from the same issue.

Patch looks good to me. Needs release notes in `docs/releases/1.8.5.txt`.

File: docs/releases/1.8.5.txt [hunk 1]
```diff
@@ -29,3 +29,6 @@ Bugfixes
 
 * Alphabetized ordering of imports in ``from django.db import migrations,
   models`` statement in newly created migrations (:ticket:`25384`).
+
+* Fixed adding a text or a blob field with unhashable default to MySQL
```
Comment (bak1an) [inline docs/releases/1.8.5.txt]: @scop Could you add the `unhashable` word to `docs/spelling_wordlist` so the docs spelling check does not fail on it? (please keep the list in alphabetical order).

File: tests/schema/tests.py [hunk 2]
```diff
@@ -1673,3 +1673,20 @@ def test_alter_field_default_dropped(self):
             )
             if connection.features.can_introspect_default:
                 self.assertIsNone(field.default)
+
+    def test_add_textfield_unhashable_default(self):
```
Comment (bak1an) [inline tests/schema/tests.py]: Tests seem to have some redundant ops.

Looks like this is enough to reproduce the issue, no need to create model instance and set `get_db_prep_save` (we are not going to save any objects.

```python
    def test_add_textfield_unhashable_default(self):
        # Create the table
        with connection.schema_editor() as editor:
            editor.create_model(Author)

        # Create a field that has an unhashable default
        new_field = TextField(default={})
        new_field.set_attributes_from_name("info")

        # Fail if an exception is raised
        with connection.schema_editor() as editor:
            editor.add_field(Author, new_field)
```

File: tests/schema/tests.py [hunk 3]
```diff
@@ -1673,3 +1673,20 @@ def test_alter_field_default_dropped(self):
...
+        # Create a field that has an unhashable default
+        new_field = TextField(default={})
+        new_field.set_attributes_from_name("info")
+
+        def get_db_prep_save(default, connection):
+            return str(default)
+        new_field.get_db_prep_save = get_db_prep_save
+        # Fail if an exception is raised
+        with connection.schema_editor() as editor:
+            editor.add_field(Author, new_field)
```
Comment (bak1an) [inline tests/schema/tests.py]: It's better to wrap this with try/except like below so when test fails we get a failure instead of error. 

```python
        # Fail if an exception is raised
        try:
            with connection.schema_editor() as editor:
                editor.add_field(Author, new_field)
        except TypeError:
            self.fail("This should not fail with TypeError")
```

Comment (MarkusH) [inline tests/schema/tests.py]: This would make more sense in a separate commit cleaning up the entire file, imo.

Comment (bak1an) [inline tests/schema/tests.py]: @MarkusH Cleaning up all the tests like this makes sense of course, but since this particular test is not yet committed and the change is rather trivial - it can be done before commit as well (and serve as an example of a good test later).

Besides, cleanup may happen but it might not as well for a plenty of reasons (no resources, unexpected difficulties, etc), so imo it's better to do this now with this particular test. Why should one commit something that already needs cleanup?

Hope this makes sense.

File: tests/schema/tests.py [hunk 4]
```diff
@@ -1673,3 +1673,19 @@ def test_alter_field_default_dropped(self):
...
+    def test_add_textfield_unhashable_default(self):
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+        # Create a row
+        Author.objects.create(name='Anonymous1')
+        # Create a field that has an unhashable default
+        new_field = TextField(default={})
+        new_field.set_attributes_from_name("info")
+        # Fail if an exception is raised
```
Comment (timgraham) [inline tests/schema/tests.py]: I don't think try/except/fail is a good pattern. See 071801ccff970682a799ce754431a3c3ce3d6902 for the reasoning.

Comment (bak1an) [inline tests/schema/tests.py]: Oops. Missed the fact that this pattern is not used in django anymore. 

@scop sorry for making you do this only to undo when core dev comes :)

**Comment by scop** [comment]
So, what is the final judgement? Does @timgraham's "I don't think [...]" mean that I need to modify the code or otherwise it will not be checked in? The related ticket is still in "ready for checkin" state and the "patch needs improvement" flag is not set. I have no idea whose word counts or who is a core dev and at which point can I trust that I can stop tweaking the code. And how do I know that if someone answers this question, that answer is from someone authoritative and something I can rely on? This is quite confusing and laborous for a trivial two character change...

From my POV it would be fair if someone would just merge this PR in its current form and make the desired adjustments on top of that in a followup commit, involving me in getting that done isn't very productive use of anyone's time.

**Comment by timgraham** [comment]
I'll amend the commit when merging, sorry for the confusion -- it happens sometimes, but we definitely appreciate your review Anton!

**Comment by timgraham** [comment]
merged in 4d933ad4181a511f3ced98edba4e17aff054e0e2, thanks!
